Source/RapyutaSimulationPlugins/Public/Core/RRActorCommon.h

```diff
@@ -0,0 +1,523 @@
+    // GameInstance.
+    friend class URRCoreUtils;
+
+private:
+    static TMap<int8, URRActorCommon*> SActorCommonList;
+
+public:
+    static constexpr int8 DEFAULT_SCENE_INSTANCE_ID = 0;
+    static URRActorCommon* GetActorCommon(int8 InSceneInstanceId = URRActorCommon::DEFAULT_SCENE_INSTANCE_ID,
+                                          UClass* ActorCommonClass = nullptr,
+                                          UObject* Outer = nullptr);
+    URRActorCommon();
+    static constexpr const TCHAR* SCRIPT_INI_PATH = TEXT("/Script/RapyutaSim.RRActorCommon");
+    static std::once_flag OnceFlag;
+
+public:
+#define EMPTY_STR (TEXT(""))    // Using TCHAR* = TEXT("") -> could causes linking error in some case!
+    static constexpr const TCHAR* SPACE_STR = TEXT(" ");
+    static constexpr const TCHAR* DELIMITER_STR = TEXT(",");
+    static constexpr const TCHAR* UNDERSCORE_STR = TEXT("_");
+    static constexpr const char* NAN_STR = "NaN";
+    static const std::string QUOTED_NAN_STR;
+    static constexpr const char* NONE_STR = "None";    // or NAME_None ?
+    static constexpr const char* NULL_STR = "Null";
+    static const std::string QUOTED_NULL_STR;
+    static constexpr const char* DOUBLE_QUOTE_CHAR = "\"";
+
+    static constexpr int32 SIM_SCENE_COMPLETION_TIMEOUT_SECS = 10;
+
+    static constexpr int8 IMAGE_BIT_DEPTH_INT8 = 8;
+    static constexpr int8 IMAGE_BIT_DEPTH_FLOAT16 = 16;
+    static constexpr int8 IMAGE_BIT_DEPTH_FLOAT32 = 32;
+
+    UPROPERTY()
+    ARRGameMode* GameMode = nullptr;
+
+    UPROPERTY()
+    ARRGameState* GameState = nullptr;
+
+    UPROPERTY()
+    int8 SceneInstanceId = DEFAULT_SCENE_INSTANCE_ID;
+
+    UPROPERTY()
+    FVector SceneInstanceLocation = FVector::ZeroVector;
+
+    virtual void PrintSimConfig() const;
+
+    virtual void OnStartSim();
+    virtual void OnBeginPlay()
+    {
+    }
+
+    virtual void OnEndPlay(const EEndPlayReason::Type EndPlayReason)
+    {
+    }
+
+    virtual void OnTick(float DeltaTime)
+    {
+    }
+
+    virtual bool HasInitialized(bool bIsLogged = false) const;
+    virtual void SetupEnvironment();
+
+    UPROPERTY()
+    TArray<AActor*> SceneLoggedActors;
+
+    UFUNCTION()
+    void ClearScene()
+    {
+    }
+
+    // UE only support custom depth stencil value in range [0-255]
+    UFUNCTION()
+    static uint8 GenerateUniqueDepthStencilValue()
+    {
+        static uint8 sLatestCustomDepthStencilValue = 0;
+        if (255 <= sLatestCustomDepthStencilValue)
+        {
+            UE_LOG(LogRapyutaCore,
+                   Error,
+                   TEXT("There are more than 255 CustomDepthStencil values having been assigned!"
+                        "Depth Segmentation Mask info would be duplicated!"));
+        }
+        return ++sLatestCustomDepthStencilValue;
+    }
+};
+
+class ARRSceneDirector;
+class ARRPlayerController;
+UCLASS()
```

Comment:
As in the PR description, this is the base class managing scene-wise operations, which will individualize world mesh loading for each scene, eg warehouse floor.
We could imagine that generically Sim could house multiple separate scenes, each instance of which has:
* `Scene director` for defining actor spawning, manipulating plan for specific scene settings/arrangement (typically customized in a synthetic data generation scene). In case of world loader, it could possibly act as the manager to define how different world instances are spawned into the scene also.
* `~Common` object handling shared operations done (like querying, housing common resources/entities, etc) in the scene.
* `~PlayerController` for camera operations in that scene only.
The idea is to maximize the throughput of a synthetic data generation app, by letting multiple scene instances run in parallel to optimize away inherent down-time in each scence's operation (like waiting for an object to settle on ground or the scene capturing itself).